File: server/python/groq_compound.py
#!/usr/bin/env python3
"""
Groq Compound AI Utility
Uses Groq's Compound model with web search, code interpreter, and visit website tools
ONLY for:
1. Realtime sentiment analysis
2. Twitter fetch AI analysis  
3. Raw data / CSV analysis

250 tokens per day limit - use sparingly!
"""

import os
import json
import logging
from groq import Groq

logging.basicConfig(level=logging.INFO)

class GroqCompoundAI:
    """
    Groq Compound AI wrapper with web search capabilities
    Limited to 250 tokens per day
    """
    
    def __init__(self):
        self.api_key = os.getenv("GROQ_API_KEY")
        if not self.api_key:
            raise ValueError("GROQ_API_KEY environment variable not found")
        
        self.client = Groq(
            api_key=self.api_key,
            default_headers={
                "Groq-Model-Version": "latest"
            }
        )
        
        self.model = "groq/compound"
        self.tools_config = {
            "tools": {
                "enabled_tools": ["web_search", "code_interpreter", "visit_website"]
            }
        }
        
        logging.info("✅ Groq Compound AI initialized (250 tokens/day limit)")
    
    def detect_language(self, text: str) -> str:
        """
        AI-based language detection using Groq
        Returns ONLY: "English", "Tagalog", or "Taglish"
        
        Args:
            text: The text to detect language for
        
        Returns:
            str: "English", "Tagalog", or "Taglish"
        """
        system_message = """You are a language detection expert for Philippines.

READ CAREFULLY: Ignore symbols, hashtags, punctuation when detecting language:
- #lindolCebu contains Tagalog word "lindol"
- (bagyo) contains Tagalog word "bagyo"  
- TULONG!!! contains Tagalog word "tulong"

Detect if the text is:
- English: Pure English text
- Tagalog: Pure Tagalog/Filipino text (includes disaster terms: bagyo, lindol, baha, sunog, ulan, tulong)
- Taglish: Mixed Tagalog and English

IMPORTANT: Respond with ONLY ONE WORD: "English", "Tagalog", or "Taglish" - nothing else!

Examples:
"omg there's a fire!" = English
"may sunog dito!" = Tagalog
"#bagyo sa Cebu" = Tagalog
"omg may sunog na naman!" = Taglish"""
        
        try:
            logging.info(f"🌐 Detecting language using AI...")
            
            completion = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "user", "content": system_message + "\n\nDetect language: " + text}
                ],
                temperature=0.1,
                max_completion_tokens=10,
                top_p=1,
                stream=False
            )
            
            language = completion.choices[0].message.content
            if not language:
                logging.warning("⚠️ AI returned empty language, defaulting to English")
                return "English"
            
            language = language.strip()
            
            # STRICT normalization - ONLY 3 values allowed: English, Tagalog, Taglish
            language_lower = language.lower()
            if "taglish" in language_lower or "mixed" in language_lower:
                return "Taglish"
            elif "tagalog" in language_lower or "filipino" in language_lower or "fil" in language_lower or language_lower == "tl":
                return "Tagalog"
            elif "english" in language_lower or language_lower == "en":
                return "English"
            else:
                # If AI returns something unexpected, default to English
                logging.warning(f"⚠️ AI returned unexpected language '{language}', defaulting to English")
                return "English"
                
        except Exception as e:
            logging.error(f"❌ Language detection error: {e}, defaulting to English")
            return "English"
    
    def analyze_sentiment_realtime(self, text: str, language: str = "English") -> dict:
        """
        Analyze sentiment for REALTIME text analysis using Compound AI
        Uses web search for recent disaster context
        
        Args:
            text: The text to analyze
            language: Detected language (English, Tagalog, Taglish)
        
        Returns:
            dict with sentiment, confidence, explanation, disasterType, location
        """
        
        # Create language-specific system message
        if language == "Tagalog":
            system_message = """Ikaw ay isang dalubhasa sa pagsusuri ng damdamin sa panahon ng sakuna sa Pilipinas.
Gamitin ang web search para sa pinakabagong konteksto ng sakuna.

CRITICAL READING INSTRUCTIONS:
1. Basahin ang mga salitang nakakabit sa symbols: #lindolCebu = "lindol" + "Cebu", (cebu) = "cebu", cebu! = "cebu"
2. Unawain ang mga emoji: 😱 = takot, 🔥 = apoy, 💪 = lakas, 🙏 = dasal
3. Basahin kahit may ALL CAPS, !!!, o maraming punctuation

FILIPINO DISASTER TERMS (Translate to English):
- bagyo = typhoon
- lindol = earthquake
- baha/pagbaha = flood
- sunog = fire
- ulan/malakas na ulan = heavy rain/storm
- guho/landslide = landslide
- aftershock = aftershock
- tsunami = tsunami

Klasipikasyon ng emosyon (sentiment):
- Panic: Matinding takot, urgency, desperasyon (TULONG!, HELP!, mamamatay na kami)
- Fear/Anxiety: Pag-aalala, kaba, takot (natatakot, kinakabahan, worried, anxious)
- Disbelief: Pagdududa, pagtataka (totoo ba?, paano nangyari?)
- Resilience: Lakas-loob, pag-asa, pagtulong (kaya natin to, tulungan, prayers)
- Neutral: Walang emosyon, factual (may lindol, may baha - walang fear words)

CRITICAL: Kahit totoo ang sakuna, kung may emotional words (natatakot, kinakabahan, worried), DAPAT Fear/Anxiety!
"Natatakot ako" = Fear/Anxiety LAGI, hindi Neutral!

IMPORTANTE: Ang "explanation" ay dapat 1 SENTENCE LANG! Maikli at malinaw.

Tumugon sa JSON format: {"sentiment": "emotion", "confidence": score, "explanation": "paliwanag", "disasterType": "English term", "location": "lokasyon"}"""
        elif language == "Taglish":
            system_message = """You are a disaster sentiment expert for Philippines analyzing Taglish (mixed Tagalog-English) messages.
Use web search for recent disaster context.

CRITICAL READING INSTRUCTIONS:
1. Read words embedded in symbols: #earthquakeCebu = "earthquake" + "Cebu", (cebu) = "cebu", cebu! = "cebu"
2. Understand emojis: 😱 = scared, 🔥 = fire, 💪 = strength, 🙏 = prayer
3. Read even with ALL CAPS, !!!, or excessive punctuation
4. Hashtags contain important info: #SaveCebu, #lindolPH

FILIPINO DISASTER TERMS (Translate to English):
- bagyo = typhoon
- lindol = earthquake
- baha/pagbaha = flood
- sunog = fire
- ulan/malakas na ulan = heavy rain/storm
- guho/landslide = landslide
- aftershock = aftershock
- tsunami = tsunami

Classify emotion (sentiment):
- Panic: Extreme fear, urgency, desperation (HELP!, RESCUE!, trapped, TULONG!)
- Fear/Anxiety: Worry, nervousness, fear (natatakot, scared, kinakabahan, worried, anxious)
- Disbelief: Doubt, shock, disbelief (totoo ba?, is this real?, paano?)
- Resilience: Strength, hope, helping others (kaya natin, prayers, tulong, support)
- Neutral: No emotion, factual (may lindol, there's a flood - no fear words)

CRITICAL: Even if disaster is real, if there are emotional words (natatakot, scared, worried), classify as Fear/Anxiety!
"Natatakot ako" or "I'm scared" = Fear/Anxiety ALWAYS, NOT Neutral!

IMPORTANT: Keep "explanation" to ONE SENTENCE ONLY! Short and clear.

Respond in JSON: {"sentiment": "emotion", "confidence": score, "explanation": "explanation", "disasterType": "English term", "location": "location"}"""
        else:
            system_message = """You are a disaster sentiment analysis expert for the Philippines.
Use web search to get recent disaster context and validate information.

CRITICAL READING INSTRUCTIONS:
1. Read words embedded in symbols/hashtags: #earthquakeCebu = "earthquake" + "Cebu", (cebu) = "cebu", cebu! = "cebu"
2. Understand emojis: 😱 = scared, 🔥 = fire, 💪 = strength, 🙏 = prayer, 😭 = crying
3. Read even with ALL CAPS, !!!, or excessive punctuation
4. Extract info from hashtags: #PrayForCebu, #TyphoonOdette, #EarthquakePH

FILIPINO DISASTER TERMS YOU MUST TRANSLATE TO ENGLISH:
- bagyo = typhoon
- lindol = earthquake  
- baha/pagbaha = flood
- sunog = fire
- ulan/malakas na ulan = heavy rain/storm
- guho/landslide = landslide
- aftershock = aftershock
- tsunami = tsunami
- pagputok ng bulkan = volcanic eruption

Classify emotional sentiment:
- Panic: Extreme fear, urgent danger, desperate help (HELP!, TRAPPED!, emergency, TULONG!)
- Fear/Anxiety: Worry, nervousness, concern (scared, afraid, worried, anxious, fearful, natatakot)
- Disbelief: Doubt, shock, can't believe it (is this real?, how did this happen?, totoo ba?)
- Resilience: Strength, hope, support, helping (we can do this, prayers, stay strong, kaya natin)
- Neutral: No emotion, factual statement (earthquake occurred, there is flooding - no fear words)

CRITICAL: Even if the disaster is real, if the person expresses fear/worry (scared, afraid, worried, natatakot), classify as Fear/Anxiety!
"I'm scared" or "I'm worried" or "Natatakot ako" = Fear/Anxiety ALWAYS, NOT Neutral!

IMPORTANT: Keep "explanation" to ONE SENTENCE ONLY! Short and clear.
IMPORTANT: Return disasterType in ENGLISH ONLY (typhoon, not bagyo; earthquake, not lindol; flood, not baha)

Respond in JSON: {"sentiment": "emotion", "confidence": score, "explanation": "explanation", "disasterType": "English term only", "location": "location"}"""
        
        try:
            logging.info(f"🤖 COMPOUND AI: Analyzing realtime text (250 tokens/day limit)")
            
            completion = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "user", "content": system_message + "\n\nAnalyze: " + text}
                ],
                temperature=0.7,
                max_completion_tokens=512,
                top_p=1,
                stream=False,
                compound_custom=self.tools_config  # type: ignore
            )
            
            response_text = completion.choices[0].message.content
            
            # Parse JSON response
            try:
                result = json.loads(response_text)
                logging.info(f"✅ COMPOUND AI: Analysis complete - {result.get('sentiment')}")
                return result
            except json.JSONDecodeError:
                # If not valid JSON, extract what we can
                logging.warning("⚠️ COMPOUND AI: Non-JSON response, parsing manually")
                return {
                    "sentiment": "Neutral",
                    "confidence": 0.5,
                    "explanation": response_text[:200],
                    "disasterType": "Unknown",
                    "location": "Unknown"
                }
                
        except Exception as e:
            logging.error(f"❌ COMPOUND AI error: {str(e)}")
            raise
    
    def analyze_csv_batch(self, texts: list, progress_callback=None) -> list:
        """
        Analyze a batch of texts for CSV processing using Compound AI
        
        Args:
            texts: List of texts to analyze
            progress_callback: Optional callback function(current, total)
        
        Returns:
            List of analysis results
        """
        results = []
        total = len(texts)
        
        logging.info(f"🤖 COMPOUND AI: Starting batch analysis of {total} texts (250 tokens/day limit)")
        
        for i, text in enumerate(texts):
            try:
                result = self.analyze_sentiment_realtime(text)
                results.append(result)
                
                if progress_callback:
                    progress_callback(i + 1, total)
                    
            except Exception as e:
                logging.error(f"❌ COMPOUND AI batch error on item {i}: {str(e)}")
                # Add fallback result
                results.append({
                    "sentiment": "Neutral",
                    "confidence": 0.3,
                    "explanation": f"Error: {str(e)}",
                    "disasterType": "Unknown",
                    "location": "Unknown"
                })
        
        logging.info(f"✅ COMPOUND AI: Batch analysis complete")
        return results


# Singleton instance
_compound_ai = None

def get_compound_ai():
    """Get or create Compound AI instance"""
    global _compound_ai
    if _compound_ai is None:
        _compound_ai = GroqCompoundAI()
    return _compound_ai

File: server/utils/groq-compound.ts
/**
 * Groq Compound AI Utility for TypeScript
 * Uses Groq's Compound model with web search, code interpreter, and visit website tools
 * ONLY for Twitter fetch AI analysis
 * 
 * 250 tokens per day limit - use sparingly!
 */

import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const GROQ_API_KEY = process.env.GROQ_API_KEY || '';
const GROQ_COMPOUND_MODEL = 'groq/compound';
const GROQ_API_URL = 'https://api.groq.com/openai/v1/chat/completions';

interface CompoundAnalysisResult {
  sentiment: string;
  confidence: number;
  language: string;
  location?: string;
  disasterType?: string;
}

export class GroqCompoundAI {
  private apiKey: string;
  
  constructor() {
    this.apiKey = GROQ_API_KEY;
    if (!this.apiKey) {
      throw new Error('GROQ_API_KEY environment variable not found');
    }
    console.log('✅ Groq Compound AI initialized for Twitter (250 tokens/day limit)');
  }
  
  /**
   * Detect language using simple heuristics
   * Returns: "English", "Tagalog", or "Taglish"
   */
  private detectLanguage(text: string): string {
    const tagalogWords = ['ang', 'mga', 'sa', 'ng', 'na', 'ay', 'ko', 'mo', 'natin', 'namin', 'kayo', 'sila', 
                          'ba', 'po', 'oo', 'hindi', 'ako', 'ikaw', 'siya', 'kami', 'tayo',
                          'may', 'wala', 'lang', 'talaga', 'pala', 'din', 'rin', 'daw',
                          'bagyo', 'lindol', 'baha', 'sunog', 'takot', 'tulong', 'kailangan'];
    
    const lowerText = text.toLowerCase();
    const words = lowerText.split(/\s+/);
    
    let tagalogCount = 0;
    let englishCount = 0;
    
    for (const word of words) {
      if (tagalogWords.includes(word)) {
        tagalogCount++;
      } else if (/^[a-z]+$/.test(word) && word.length > 2) {
        englishCount++;
      }
    }
    
    // If both languages present, it's Taglish
    if (tagalogCount > 0 && englishCount > 0) {
      return 'Taglish';
    }
    
    // If mostly Tagalog
    if (tagalogCount > englishCount) {
      return 'Tagalog';
    }
    
    // Default to English
    return 'English';
  }
  
  /**
   * Analyze tweet sentiment using Compound AI with web search
   * SAME ALGORITHM as realtime analysis but WITHOUT explanation
   * @param text - The tweet text to analyze
   * @returns Analysis result with sentiment, confidence, location, disaster type
   */
  async analyzeTweetSentiment(text: string): Promise<CompoundAnalysisResult> {
    // Detect language first
    const language = this.detectLanguage(text);
    
    // Use EXACT SAME prompts as Python realtime analysis, but WITHOUT explanation in response
    let systemPrompt: string;
    
    if (language === 'Tagalog') {
      systemPrompt = `Ikaw ay isang dalubhasa sa pagsusuri ng damdamin sa panahon ng sakuna sa Pilipinas.
Gamitin ang web search para sa pinakabagong konteksto ng sakuna.

FILIPINO DISASTER TERMS - ONLY THESE 5 TYPES:
- bagyo = Typhoon
- lindol = Earthquake
- sunog = Fire
- guho/landslide = Landslide
- bulkan/pagputok = Volcanic Eruption

IMPORTANT: Only classify these 5 disaster types. Ignore flood, tsunami, storm, or others.

Klasipikasyon ng emosyon (sentiment):
- Panic: Matinding takot, urgency, desperasyon (TULONG!, HELP!, mamamatay na kami)
- Fear/Anxiety: Pag-aalala, kaba, takot (natatakot, kinakabahan, worried, anxious)
- Disbelief: Pagdududa, pagtataka (totoo ba?, paano nangyari?)
- Resilience: Lakas-loob, pag-asa, pagtulong (kaya natin to, tulungan, prayers)
- Neutral: Walang emosyon, factual (may lindol, may baha - walang fear words)

CRITICAL: Kahit totoo ang sakuna, kung may emotional words (natatakot, kinakabahan, worried), DAPAT Fear/Anxiety!
"Natatakot ako" = Fear/Anxiety LAGI, hindi Neutral!

Tumugon sa JSON format: {"sentiment": "emotion", "confidence": score, "disasterType": "Typhoon/Fire/Volcanic Eruption/Earthquake/Landslide ONLY", "location": "lokasyon"}`;
    } else if (language === 'Taglish') {
      systemPrompt = `You are a disaster sentiment expert for Philippines analyzing Taglish (mixed Tagalog-English) messages.
Use web search for recent disaster context.

FILIPINO DISASTER TERMS - ONLY THESE 5 TYPES:
- bagyo = Typhoon
- lindol = Earthquake
- sunog = Fire
- guho/landslide = Landslide
- bulkan/pagputok = Volcanic Eruption

IMPORTANT: Only classify these 5 disaster types. Ignore flood, tsunami, storm, or others.

Classify emotion (sentiment):
- Panic: Extreme fear, urgency, desperation (HELP!, RESCUE!, trapped)
- Fear/Anxiety: Worry, nervousness, fear (natatakot, scared, kinakabahan, worried, anxious)
- Disbelief: Doubt, shock, disbelief (totoo ba?, is this real?, paano?)
- Resilience: Strength, hope, helping others (kaya natin, prayers, tulong, support)
- Neutral: No emotion, factual (may lindol, there's a flood - no fear words)

CRITICAL: Even if disaster is real, if there are emotional words (natatakot, scared, worried), classify as Fear/Anxiety!
"Natatakot ako" or "I'm scared" = Fear/Anxiety ALWAYS, NOT Neutral!

Respond in JSON: {"sentiment": "emotion", "confidence": score, "disasterType": "Typhoon/Fire/Volcanic Eruption/Earthquake/Landslide ONLY", "location": "location"}`;
    } else {
      systemPrompt = `You are a disaster sentiment analysis expert for the Philippines.
Use web search to get recent disaster context and validate information.

FILIPINO DISASTER TERMS - ONLY THESE 5 TYPES:
- bagyo = Typhoon
- lindol = Earthquake
- sunog = Fire
- guho/landslide = Landslide
- bulkan/pagputok ng bulkan = Volcanic Eruption

IMPORTANT: Only classify these 5 disaster types. Ignore flood, baha, tsunami, storm, drought, or any others.

Classify emotional sentiment:
- Panic: Extreme fear, urgent danger, desperate help (HELP!, TRAPPED!, emergency)
- Fear/Anxiety: Worry, nervousness, concern (scared, afraid, worried, anxious, fearful)
- Disbelief: Doubt, shock, can't believe it (is this real?, how did this happen?)
- Resilience: Strength, hope, support, helping (we can do this, prayers, stay strong)
- Neutral: No emotion, factual statement (earthquake occurred, there is flooding - no fear words)

CRITICAL: Even if the disaster is real, if the person expresses fear/worry (scared, afraid, worried), classify as Fear/Anxiety!
"I'm scared" or "I'm worried" = Fear/Anxiety ALWAYS, NOT Neutral!

IMPORTANT: Return disasterType with proper capitalization (Typhoon, Fire, Volcanic Eruption, Earthquake, or Landslide ONLY).

Respond in JSON: {"sentiment": "emotion", "confidence": score, "disasterType": "Typhoon/Fire/Volcanic Eruption/Earthquake/Landslide ONLY", "location": "location"}`;
    }

    try {
      console.log(`🤖 COMPOUND AI (Twitter): Analyzing tweet as ${language} (250 tokens/day limit)`);
      
      const response = await axios.post(
        GROQ_API_URL,
        {
          model: GROQ_COMPOUND_MODEL,
          messages: [
            { role: 'user', content: systemPrompt + '\n\nAnalyze: ' + text }
          ],
          temperature: 0.7,
          max_tokens: 512,
          top_p: 1,
          stream: false,
          // Groq Compound specific configuration
          compound_custom: {
            tools: {
              enabled_tools: ['web_search', 'code_interpreter', 'visit_website']
            }
          }
        },
        {
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json',
            'Groq-Model-Version': 'latest'
          },
          timeout: 30000
        }
      );
      
      const content = response.data.choices[0].message.content;
      
      // Parse JSON response
      try {
        const result = JSON.parse(content);
        
        // Set the detected language
        result.language = language;
        
        console.log(`✅ COMPOUND AI (Twitter): Analysis complete - ${result.sentiment}, Language: ${result.language}`);
        return result;
      } catch (parseError) {
        console.warn('⚠️ COMPOUND AI (Twitter): Non-JSON response, parsing manually');
        return {
          sentiment: 'Neutral',
          confidence: 0.5,
          language: language
        };
      }
      
    } catch (error: any) {
      console.error('❌ COMPOUND AI (Twitter) error:', error.message);
      throw error;
    }
  }
}

// Singleton instance
let compoundAI: GroqCompoundAI | null = null;

export function getCompoundAI(): GroqCompoundAI {
  if (!compoundAI) {
    compoundAI = new GroqCompoundAI();
  }
  return compoundAI;
}
